File: libraries/python/src/pandaprobe/integrations/openai_agents/adapter.py
"""OpenAI Agents SDK integration adapter for PandaProbe tracing.

Instruments the OpenAI Agents SDK by implementing its first-class
``TracingProcessor`` callback interface.  Each ``Runner.run`` invocation
produces a full ``TraceData`` with nested ``SpanData`` objects for agents,
LLM calls (Responses API and Chat Completions), tool invocations, handoffs,
and guardrails.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from pandaprobe.integrations._base import BaseIntegrationAdapter, safe_serialize
from pandaprobe.integrations.openai_agents.utils import (
    extract_generation_model_parameters,
    extract_reasoning,
    extract_response_model_parameters,
    extract_token_usage,
    normalize_generation_input,
    normalize_generation_output,
    normalize_response_input,
    normalize_response_output,
    serialize_tool_io,
)
from pandaprobe.schemas import SpanData, SpanKind, SpanStatusCode, TraceData, TraceStatus
from pandaprobe.tracing.session import get_current_session_id, get_current_user_id
from pandaprobe.validation import extract_last_user_message

logger = logging.getLogger("pandaprobe")

_patched = False

# ---------------------------------------------------------------------------
# Per-invocation trace state
# ---------------------------------------------------------------------------

_trace_states: dict[str, _TraceState] = {}


@dataclass
class _TraceState:
    """Per-invocation state accumulated during a single Runner.run call."""

    adapter: OpenAIAgentsAdapter
    spans: dict[str, SpanData] = field(default_factory=dict)
    root_span_id: str | None = None
    trace_input: Any = None
    trace_output: Any = None
    chain_input: Any = None
    chain_output: Any = None
    sdk_to_pandaprobe: dict[str, str] = field(default_factory=dict)
    trace_id: str | None = None
    trace_started_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


# ---------------------------------------------------------------------------
# Adapter
# ---------------------------------------------------------------------------


class OpenAIAgentsAdapter(BaseIntegrationAdapter):
    """PandaProbe integration adapter for the OpenAI Agents SDK.

    Usage::

        from pandaprobe.integrations.openai_agents import OpenAIAgentsAdapter

        adapter = OpenAIAgentsAdapter(tags=["example"])
        adapter.instrument()

        # All subsequent Runner.run() calls are now traced automatically.
    """

    def instrument(self) -> bool:
        """Register a ``TracingProcessor`` with the OpenAI Agents SDK.

        Returns ``True`` if registration succeeded, ``False`` if dependencies
        are missing.  The method is idempotent — calling it multiple times is
        safe.
        """
        global _patched

        _store_adapter(self)

        if _patched:
            return True

        try:
            from agents import tracing as agents_tracing  # type: ignore[import-not-found]
        except ImportError as exc:
            logger.warning("PandaProbe OpenAI Agents SDK integration: missing dependency — %s", exc)
            return False

        class _PandaProbeProcessor(agents_tracing.TracingProcessor):
            """Bridge between OpenAI Agents SDK tracing and PandaProbe."""

            def on_trace_start(self, trace: Any) -> None:
                _on_trace_start(trace)

            def on_trace_end(self, trace: Any) -> None:
                _on_trace_end(trace)

            def on_span_start(self, span: Any) -> None:
                _on_span_start(span)

            def on_span_end(self, span: Any) -> None:
                _on_span_end(span)

            def shutdown(self) -> None:
                pass

            def force_flush(self) -> None:
                pass

        agents_tracing.add_trace_processor(_PandaProbeProcessor())

        _patched = True
        return True

    # ------------------------------------------------------------------
    # Finalization
    # ------------------------------------------------------------------

    def _finalize_trace(self, state: _TraceState, *, error: bool = False) -> None:
        """Assemble a TraceData from collected spans and submit it."""
        try:
            client = self._resolve_client()
            spans = list(state.spans.values())
            session_id = self._session_id if self._session_id is not None else get_current_session_id()
            user_id = self._user_id if self._user_id is not None else get_current_user_id()

            trace = TraceData(
                name="OpenAIAgentsSDK",
                status=TraceStatus.ERROR if error else TraceStatus.COMPLETED,
                input=state.trace_input,
                output=state.trace_output,
                metadata=dict(self._metadata),
                started_at=state.trace_started_at,
                ended_at=datetime.now(timezone.utc),
                session_id=session_id,
                user_id=user_id,
                tags=list(self._tags),
                spans=spans,
            )
            client.log_trace(trace)
        except Exception as exc:
            logger.error("PandaProbe OpenAI Agents SDK adapter failed to submit trace: %s", exc)


# ---------------------------------------------------------------------------
# Module-level adapter storage
# ---------------------------------------------------------------------------

_active_adapter: OpenAIAgentsAdapter | None = None


def _store_adapter(adapter: OpenAIAgentsAdapter) -> None:
    global _active_adapter
    _active_adapter = adapter


def _get_adapter() -> OpenAIAgentsAdapter | None:
    return _active_adapter


# ---------------------------------------------------------------------------
# Span kind resolution
# ---------------------------------------------------------------------------

_SPAN_TYPE_TO_KIND: dict[str, SpanKind] = {
    "agent": SpanKind.AGENT,
    "handoff": SpanKind.AGENT,
    "response": SpanKind.LLM,
    "generation": SpanKind.LLM,
    "function": SpanKind.TOOL,
    "guardrail": SpanKind.OTHER,
    "custom": SpanKind.OTHER,
}


def _resolve_span_kind(span_data: Any) -> SpanKind:
    """Determine the PandaProbe SpanKind from an SDK span_data object."""
    span_type = getattr(span_data, "type", None)
    if span_type and isinstance(span_type, str):
        return _SPAN_TYPE_TO_KIND.get(span_type, SpanKind.OTHER)
    return SpanKind.OTHER


def _resolve_span_name(span: Any) -> str:
    """Derive a human-readable name from an SDK span."""
    span_data = getattr(span, "span_data", None)
    if span_data is not None:
        name = getattr(span_data, "name", None)
        if name:
            return str(name)

    name = getattr(span, "name", None)
    if name:
        return str(name)

    span_type = getattr(span_data, "type", None) if span_data else None
    if span_type:
        return str(span_type).capitalize()

    return "Span"


# ---------------------------------------------------------------------------
# TracingProcessor event handlers
# ---------------------------------------------------------------------------


def _on_trace_start(trace: Any) -> None:
    """Handle trace start: create _TraceState and root CHAIN span."""
    adapter = _get_adapter()
    if adapter is None:
        return

    try:
        trace_id = getattr(trace, "trace_id", None)
        if trace_id is None:
            return

        trace_id = str(trace_id)

        state = _TraceState(adapter=adapter, trace_id=trace_id)

        root_span_id = str(uuid4())
        state.root_span_id = root_span_id

        trace_name = getattr(trace, "name", None) or "OpenAIAgentsSDK"

        root_span = SpanData(
            span_id=root_span_id,
            parent_span_id=None,
            name=str(trace_name),
            kind=SpanKind.CHAIN,
            started_at=state.trace_started_at,
        )
        state.spans[root_span_id] = root_span
        _trace_states[trace_id] = state
    except Exception as exc:
        logger.debug("PandaProbe OpenAI Agents: error in on_trace_start — %s", exc)


def _on_trace_end(trace: Any) -> None:
    """Handle trace end: finalize root CHAIN span and submit trace."""
    try:
        trace_id = str(getattr(trace, "trace_id", ""))
        state = _trace_states.pop(trace_id, None)
        if state is None:
            return

        now = datetime.now(timezone.utc)
        root_span = state.spans.get(state.root_span_id) if state.root_span_id else None

        has_error = any(s.status == SpanStatusCode.ERROR for s in state.spans.values())

        if root_span is not None:
            root_span.input = state.chain_input
            root_span.output = state.chain_output
            root_span.ended_at = now
            root_span.status = SpanStatusCode.ERROR if has_error else SpanStatusCode.OK

        state.adapter._finalize_trace(state, error=has_error)
    except Exception as exc:
        logger.debug("PandaProbe OpenAI Agents: error in on_trace_end — %s", exc)


def _on_span_start(span: Any) -> None:
    """Handle span start: create SpanData, resolve parent, register ID mapping."""
    try:
        trace_id = str(getattr(span, "trace_id", ""))
        state = _trace_states.get(trace_id)
        if state is None:
            return

        sdk_span_id = str(getattr(span, "span_id", ""))
        if not sdk_span_id:
            return

        span_data = getattr(span, "span_data", None)
        kind = _resolve_span_kind(span_data)
        name = _resolve_span_name(span)

        pp_span_id = str(uuid4())
        state.sdk_to_pandaprobe[sdk_span_id] = pp_span_id

        sdk_parent_id = getattr(span, "parent_id", None)
        if sdk_parent_id is not None:
            sdk_parent_id = str(sdk_parent_id)

        if sdk_parent_id is None or sdk_parent_id == trace_id:
            parent_pp_id = state.root_span_id
        else:
            parent_pp_id = state.sdk_to_pandaprobe.get(sdk_parent_id, state.root_span_id)

        started_at = None
        raw_started = getattr(span, "started_at", None)
        if raw_started:
            try:
                started_at = datetime.fromisoformat(str(raw_started))
            except (ValueError, TypeError):
                pass
        if started_at is None:
            started_at = datetime.now(timezone.utc)

        pp_span = SpanData(
            span_id=pp_span_id,
            parent_span_id=parent_pp_id,
            name=name,
            kind=kind,
            started_at=started_at,
        )
        state.spans[pp_span_id] = pp_span
    except Exception as exc:
        logger.debug("PandaProbe OpenAI Agents: error in on_span_start — %s", exc)


def _on_span_end(span: Any) -> None:
    """Handle span end: fill in output, token usage, model params, reasoning, error."""
    try:
        trace_id = str(getattr(span, "trace_id", ""))
        state = _trace_states.get(trace_id)
        if state is None:
            return

        sdk_span_id = str(getattr(span, "span_id", ""))
        pp_span_id = state.sdk_to_pandaprobe.get(sdk_span_id)
        if pp_span_id is None:
            return

        pp_span = state.spans.get(pp_span_id)
        if pp_span is None:
            return

        ended_at = None
        raw_ended = getattr(span, "ended_at", None)
        if raw_ended:
            try:
                ended_at = datetime.fromisoformat(str(raw_ended))
            except (ValueError, TypeError):
                pass
        if ended_at is None:
            ended_at = datetime.now(timezone.utc)
        pp_span.ended_at = ended_at

        error = getattr(span, "error", None)
        if error:
            pp_span.error = str(error)
            pp_span.status = SpanStatusCode.ERROR
        else:
            pp_span.status = SpanStatusCode.OK

        span_data = getattr(span, "span_data", None)
        if span_data is None:
            return

        span_type = getattr(span_data, "type", None)

        if span_type == "response":
            _fill_response_span(pp_span, span_data, state)
        elif span_type == "generation":
            _fill_generation_span(pp_span, span_data, state)
        elif span_type == "function":
            _fill_function_span(pp_span, span_data)
        elif span_type == "agent":
            _fill_agent_span(pp_span, span_data)
        elif span_type == "handoff":
            _fill_handoff_span(pp_span, span_data)
        elif span_type == "guardrail":
            _fill_guardrail_span(pp_span, span_data)
        elif span_type == "custom":
            _fill_custom_span(pp_span, span_data)

    except Exception as exc:
        logger.debug("PandaProbe OpenAI Agents: error in on_span_end — %s", exc)


# ---------------------------------------------------------------------------
# Span-type-specific data extraction
# ---------------------------------------------------------------------------


def _propagate_llm_to_parent_agent(pp_span: SpanData, state: _TraceState) -> None:
    """Propagate an LLM span's input/output to its parent AGENT span.

    The first child LLM span's input becomes the agent's input.
    The last child LLM span's output always updates the agent's output.
    """
    parent_id = str(pp_span.parent_span_id) if pp_span.parent_span_id else None
    if not parent_id:
        return
    parent_span = state.spans.get(parent_id)
    if parent_span is None or parent_span.kind != SpanKind.AGENT:
        return
    if parent_span.input is None and pp_span.input:
        parent_span.input = pp_span.input
    if pp_span.output:
        parent_span.output = pp_span.output


def _update_chain_and_trace_io(pp_span: SpanData, state: _TraceState) -> None:
    """Update chain-level and trace-level I/O from an LLM span.

    Chain I/O: the first LLM span's full input and a combined
    (input messages + output messages) for output, giving the full
    conversation including the new assistant response.

    Trace I/O: trimmed — last user message for input, last assistant
    output for output.
    """
    if state.chain_input is None and pp_span.input:
        state.chain_input = pp_span.input

    chain_messages: list[dict[str, Any]] = []
    if pp_span.input and isinstance(pp_span.input, dict):
        chain_messages.extend(pp_span.input.get("messages", []))
    if pp_span.output and isinstance(pp_span.output, dict):
        chain_messages.extend(pp_span.output.get("messages", []))
    if chain_messages:
        state.chain_output = {"messages": chain_messages}
    elif pp_span.output:
        state.chain_output = pp_span.output

    if state.trace_input is None and pp_span.input:
        state.trace_input = extract_last_user_message(pp_span.input)
    if pp_span.output:
        state.trace_output = pp_span.output


def _fill_response_span(pp_span: SpanData, span_data: Any, state: _TraceState) -> None:
    """Extract rich data from a ResponseSpanData."""
    pp_span.input = normalize_response_input(span_data)
    pp_span.output = normalize_response_output(span_data)

    response = getattr(span_data, "response", None)
    if response is not None:
        model = getattr(response, "model", None)
        if model:
            pp_span.model = str(model)
            pp_span.name = str(model)

        usage = getattr(response, "usage", None)
        if usage:
            pp_span.token_usage = extract_token_usage(usage)

        pp_span.model_parameters = extract_response_model_parameters(response)

        reasoning = extract_reasoning(response)
        if reasoning:
            pp_span.metadata["reasoning_summary"] = reasoning

    _propagate_llm_to_parent_agent(pp_span, state)
    _update_chain_and_trace_io(pp_span, state)


def _fill_generation_span(pp_span: SpanData, span_data: Any, state: _TraceState) -> None:
    """Extract data from a GenerationSpanData."""
    pp_span.input = normalize_generation_input(span_data)
    pp_span.output = normalize_generation_output(span_data)

    model = getattr(span_data, "model", None)
    if model:
        pp_span.model = str(model)
        pp_span.name = str(model)

    usage = getattr(span_data, "usage", None)
    if usage:
        pp_span.token_usage = extract_token_usage(usage)

    pp_span.model_parameters = extract_generation_model_parameters(span_data)

    _propagate_llm_to_parent_agent(pp_span, state)
    _update_chain_and_trace_io(pp_span, state)


def _fill_function_span(pp_span: SpanData, span_data: Any) -> None:
    """Extract data from a FunctionSpanData."""
    raw_input = getattr(span_data, "input", None)
    if raw_input is not None:
        pp_span.input = serialize_tool_io(raw_input)

    raw_output = getattr(span_data, "output", None)
    if raw_output is not None:
        pp_span.output = serialize_tool_io(raw_output)

    name = getattr(span_data, "name", None)
    if name:
        pp_span.name = str(name)

    mcp_data = getattr(span_data, "mcp_data", None)
    if mcp_data:
        pp_span.metadata["mcp_data"] = safe_serialize(mcp_data)


def _fill_agent_span(pp_span: SpanData, span_data: Any) -> None:
    """Extract metadata from an AgentSpanData."""
    tools = getattr(span_data, "tools", None)
    if tools:
        pp_span.metadata["tools"] = safe_serialize(tools)

    handoffs = getattr(span_data, "handoffs", None)
    if handoffs:
        pp_span.metadata["handoffs"] = safe_serialize(handoffs)

    output_type = getattr(span_data, "output_type", None)
    if output_type:
        pp_span.metadata["output_type"] = str(output_type)


def _fill_handoff_span(pp_span: SpanData, span_data: Any) -> None:
    """Extract data from a HandoffSpanData."""
    from_agent = getattr(span_data, "from_agent", None)
    to_agent = getattr(span_data, "to_agent", None)

    pp_span.input = {"from_agent": str(from_agent) if from_agent else None}
    pp_span.output = {"to_agent": str(to_agent) if to_agent else None}
    pp_span.metadata["from_agent"] = str(from_agent) if from_agent else None
    pp_span.metadata["to_agent"] = str(to_agent) if to_agent else None


def _fill_guardrail_span(pp_span: SpanData, span_data: Any) -> None:
    """Extract data from a GuardrailSpanData."""
    triggered = getattr(span_data, "triggered", None)
    pp_span.metadata["triggered"] = triggered


def _fill_custom_span(pp_span: SpanData, span_data: Any) -> None:
    """Extract data from a CustomSpanData."""
    data = getattr(span_data, "data", None)
    if data:
        pp_span.metadata["custom_data"] = safe_serialize(data)

File: libraries/typescript/src/integrations/openai-agents/adapter.ts
/**
 * OpenAI Agents SDK integration.
 *
 * Mirrors the Python adapter: register a tracing processor with the
 * `@openai/agents` SDK and build, per SDK trace, a synthetic root CHAIN span
 * that parents all SDK spans (remapped to PandaProbe UUIDs). Each SDK span is
 * filled per type (response/generation/function/agent/handoff/guardrail/custom)
 * with normalized message I/O, token usage, model parameters and reasoning; LLM
 * spans propagate I/O to their parent AGENT span and drive chain/trace-level I/O.
 */

import { logger } from "../../logger.js";
import { SpanData, SpanKind, SpanStatusCode, TraceData, TraceStatus, newUuid } from "../../schemas.js";
import { getCurrentSessionId, getCurrentUserId } from "../../tracing/session.js";
import { extractLastUserMessage } from "../../validation.js";
import { type AdapterOptions, BaseIntegrationAdapter, safeSerialize } from "../base.js";
import {
  extractGenerationModelParameters,
  extractReasoning,
  extractResponseModelParameters,
  extractTokenUsage,
  mapSpanKind,
  normalizeGenerationInput,
  normalizeGenerationOutput,
  normalizeResponseInput,
  normalizeResponseOutput,
  resolveSpanName,
  serializeToolIo,
} from "./utils.js";

type Any = any;

const TRACE_NAME = "OpenAIAgentsSDK";

interface TraceState {
  rootSpanId: string;
  startedAt: Date;
  spans: Map<string, SpanData>;
  sdkToPp: Map<string, string>;
  traceInput: unknown;
  traceOutput: unknown;
  chainInput: unknown;
  chainOutput: unknown;
}

/** The TracingProcessor object shape expected by `@openai/agents`. */
export interface TraceProcessor {
  onTraceStart(trace: Any): void;
  onTraceEnd(trace: Any): void;
  onSpanStart(span: Any): void;
  onSpanEnd(span: Any): void;
  forceFlush?(): void;
  shutdown?(): void;
}

function messagesOf(io: unknown): Any[] {
  if (io && typeof io === "object" && Array.isArray((io as Any).messages)) {
    return (io as Any).messages;
  }
  return [];
}

export class OpenAIAgentsAdapter extends BaseIntegrationAdapter {
  private states = new Map<string, TraceState>();

  /** Build the tracing processor object (register it with the Agents SDK). */
  createTraceProcessor(): TraceProcessor {
    return {
      onTraceStart: (trace: Any) => {
        const traceId = String(trace?.traceId ?? trace?.trace_id ?? "");
        if (!traceId) {
          return;
        }
        const rootSpanId = newUuid();
        const startedAt = new Date();
        const spans = new Map<string, SpanData>();
        spans.set(
          rootSpanId,
          new SpanData({
            spanId: rootSpanId,
            parentSpanId: null,
            name: trace?.name ?? TRACE_NAME,
            kind: SpanKind.CHAIN,
            startedAt,
          }),
        );
        this.states.set(traceId, {
          rootSpanId,
          startedAt,
          spans,
          sdkToPp: new Map(),
          traceInput: null,
          traceOutput: null,
          chainInput: null,
          chainOutput: null,
        });
      },

      onSpanStart: (span: Any) => {
        const traceId = String(span?.traceId ?? span?.trace_id ?? "");
        const state = this.states.get(traceId);
        if (!state) {
          return;
        }
        const sdkSpanId = String(span?.spanId ?? span?.span_id ?? "");
        if (!sdkSpanId) {
          return;
        }
        const spanData = span?.spanData ?? span?.span_data;
        const ppId = newUuid();
        state.sdkToPp.set(sdkSpanId, ppId);

        const rawParent = span?.parentId ?? span?.parent_id;
        const sdkParent = rawParent != null ? String(rawParent) : null;
        const parentPp =
          sdkParent === null || sdkParent === traceId
            ? state.rootSpanId
            : (state.sdkToPp.get(sdkParent) ?? state.rootSpanId);

        state.spans.set(
          ppId,
          new SpanData({
            spanId: ppId,
            parentSpanId: parentPp,
            name: resolveSpanName(span),
            kind: mapSpanKind(spanData?.type),
            startedAt: span?.startedAt ? new Date(span.startedAt) : new Date(),
          }),
        );
      },

      onSpanEnd: (span: Any) => {
        const traceId = String(span?.traceId ?? span?.trace_id ?? "");
        const state = this.states.get(traceId);
        if (!state) {
          return;
        }
        const sdkSpanId = String(span?.spanId ?? span?.span_id ?? "");
        const ppId = state.sdkToPp.get(sdkSpanId);
        const sd = ppId ? state.spans.get(ppId) : undefined;
        if (!sd) {
          return;
        }
        sd.endedAt = span?.endedAt ? new Date(span.endedAt) : new Date();
        if (span?.error) {
          sd.error = String(span.error?.message ?? span.error);
          sd.status = SpanStatusCode.ERROR;
        } else {
          sd.status = SpanStatusCode.OK;
        }

        const spanData = span?.spanData ?? span?.span_data;
        if (!spanData) {
          return;
        }
        switch (spanData.type) {
          case "response":
            this.fillResponseSpan(sd, spanData, state);
            break;
          case "generation":
            this.fillGenerationSpan(sd, spanData, state);
            break;
          case "function":
            this.fillFunctionSpan(sd, spanData);
            break;
          case "agent":
            this.fillAgentSpan(sd, spanData);
            break;
          case "handoff":
            this.fillHandoffSpan(sd, spanData);
            break;
          case "guardrail":
            sd.metadata.triggered = spanData.triggered;
            break;
          case "custom":
            if (spanData.data) {
              sd.metadata.custom_data = safeSerialize(spanData.data);
            }
            break;
        }
      },

      onTraceEnd: (trace: Any) => {
        const traceId = String(trace?.traceId ?? trace?.trace_id ?? "");
        const state = this.states.get(traceId);
        if (!state) {
          return;
        }
        this.states.delete(traceId);
        try {
          const now = new Date();
          const hasError = [...state.spans.values()].some((s) => s.status === SpanStatusCode.ERROR);
          const root = state.spans.get(state.rootSpanId);
          if (root) {
            root.input = state.chainInput;
            root.output = state.chainOutput;
            root.endedAt = now;
            root.status = hasError ? SpanStatusCode.ERROR : SpanStatusCode.OK;
          }
          this.resolveClient().logTrace(
            new TraceData({
              name: TRACE_NAME,
              status: hasError ? TraceStatus.ERROR : TraceStatus.COMPLETED,
              input: state.traceInput,
              output: state.traceOutput,
              metadata: { ...this.metadata },
              startedAt: state.startedAt,
              endedAt: now,
              sessionId: this.sessionId ?? getCurrentSessionId(),
              userId: this.userId ?? getCurrentUserId(),
              tags: [...this.tags],
              spans: [...state.spans.values()],
            }),
          );
        } catch (exc) {
          logger.error(`OpenAIAgents adapter failed to submit trace: ${String(exc)}`);
        }
      },

      forceFlush: () => {},
      shutdown: () => {},
    };
  }

  // ------------------------------------------------------------------
  // Per-span-type fills
  // ------------------------------------------------------------------

  private propagateLlmToParentAgent(span: SpanData, state: TraceState): void {
    const parentId = span.parentSpanId;
    if (!parentId) {
      return;
    }
    const parent = state.spans.get(parentId);
    if (!parent || parent.kind !== SpanKind.AGENT) {
      return;
    }
    if ((parent.input === null || parent.input === undefined) && span.input) {
      parent.input = span.input;
    }
    if (span.output) {
      parent.output = span.output;
    }
  }

  private updateChainAndTraceIo(span: SpanData, state: TraceState): void {
    if ((state.chainInput === null || state.chainInput === undefined) && span.input) {
      state.chainInput = span.input;
    }
    const chainMessages = [...messagesOf(span.input), ...messagesOf(span.output)];
    if (chainMessages.length > 0) {
      state.chainOutput = { messages: chainMessages };
    } else if (span.output) {
      state.chainOutput = span.output;
    }
    if ((state.traceInput === null || state.traceInput === undefined) && span.input) {
      state.traceInput = extractLastUserMessage(span.input);
    }
    if (span.output) {
      state.traceOutput = span.output;
    }
  }

  private fillResponseSpan(span: SpanData, spanData: Any, state: TraceState): void {
    span.input = normalizeResponseInput(spanData);
    span.output = normalizeResponseOutput(spanData);
    const response = spanData.response;
    if (response) {
      if (response.model) {
        span.model = String(response.model);
        span.name = String(response.model);
      }
      if (response.usage) {
        span.tokenUsage = extractTokenUsage(response.usage);
      }
      span.modelParameters = extractResponseModelParameters(response);
      const reasoning = extractReasoning(response);
      if (reasoning) {
        span.metadata.reasoning_summary = reasoning;
      }
    }
    this.propagateLlmToParentAgent(span, state);
    this.updateChainAndTraceIo(span, state);
  }

  private fillGenerationSpan(span: SpanData, spanData: Any, state: TraceState): void {
    span.input = normalizeGenerationInput(spanData);
    span.output = normalizeGenerationOutput(spanData);
    if (spanData.model) {
      span.model = String(spanData.model);
      span.name = String(spanData.model);
    }
    if (spanData.usage) {
      span.tokenUsage = extractTokenUsage(spanData.usage);
    }
    span.modelParameters = extractGenerationModelParameters(spanData);
    this.propagateLlmToParentAgent(span, state);
    this.updateChainAndTraceIo(span, state);
  }

  private fillFunctionSpan(span: SpanData, spanData: Any): void {
    if (spanData.input != null) {
      span.input = serializeToolIo(spanData.input);
    }
    if (spanData.output != null) {
      span.output = serializeToolIo(spanData.output);
    }
    if (spanData.name) {
      span.name = String(spanData.name);
    }
    if (spanData.mcp_data) {
      span.metadata.mcp_data = safeSerialize(spanData.mcp_data);
    }
  }

  private fillAgentSpan(span: SpanData, spanData: Any): void {
    if (spanData.tools) {
      span.metadata.tools = safeSerialize(spanData.tools);
    }
    if (spanData.handoffs) {
      span.metadata.handoffs = safeSerialize(spanData.handoffs);
    }
    if (spanData.output_type) {
      span.metadata.output_type = String(spanData.output_type);
    }
  }

  private fillHandoffSpan(span: SpanData, spanData: Any): void {
    const from = spanData.from_agent ? String(spanData.from_agent) : null;
    const to = spanData.to_agent ? String(spanData.to_agent) : null;
    span.input = { from_agent: from };
    span.output = { to_agent: to };
    span.metadata.from_agent = from;
    span.metadata.to_agent = to;
  }

  /**
   * Register the processor with the `@openai/agents` SDK. Returns true on
   * success, false if the SDK is unavailable or exposes no known hook.
   */
  async instrument(): Promise<boolean> {
    let agents: Any;
    try {
      // Indirect the specifier so the bundler/typechecker treats it as a
      // runtime-only optional import (the SDK is an optional peer dependency).
      const moduleName: string = "@openai/agents";
      agents = await import(moduleName);
    } catch {
      return false;
    }
    const processor = this.createTraceProcessor();
    if (typeof agents.addTraceProcessor === "function") {
      agents.addTraceProcessor(processor);
      return true;
    }
    const tracing = agents.tracing ?? agents.getGlobalTraceProvider?.();
    if (tracing && typeof tracing.registerProcessor === "function") {
      tracing.registerProcessor(processor);
      return true;
    }
    return false;
  }
}
